### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by Page-**
Change-type: major

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    "lint-fix": "balena-lint --typescript --fix src test",
+    "test": "mocha && npm run lint",
+    "prepack": "npm run build",
+    "prepare": "husky install"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/balena-io-modules/balena-env-parsing.git"
+  },
+  "author": "Page-",
```
Comment (dfunckt) [inline package.json:20]: Balena Ltd.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": "https://github.com/balena-io-modules/balena-env-parsing/issues"
+  },
+  "homepage": "https://github.com/balena-io-modules/balena-env-parsing#readme",
```
Comment (dfunckt) [inline package.json:25]:
```suggestion
  "homepage": "https://github.com/balena-io-modules/balena-env-parsing",
```

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	}
+	const i = parseInt(s, 10);
+	if (!Number.isFinite(i)) {
+		return;
+	}
+	return i;
+};
+
+/**
+ * Fetches the contents of an env var, ensuring it is a valid integer if set, if it is not set then instead it returns the specified default value
```
Comment (dfunckt) [inline src/index.ts:54]:
```suggestion
 * Fetches the contents of an env var, ensuring it is a valid integer if set,
 * if it is not set then instead it returns the specified default value.
```

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+	const i = checkInt(s);
+	if (i === undefined) {
+		throw new Error(`${varName} must be a valid number if set`);
+	}
+	return i;
+}
+
+/**
+ * Fetches the contents of an env var, ensuring it is a valid boolean (`'true'`/`'false'`) if set, if it is not set then instead it returns the specified default value
```
Comment (dfunckt) [inline src/index.ts:76]:
```suggestion
 * Fetches the contents of an env var, ensuring it is a valid boolean (`'true'`/`'false'`) if set,
 * if it is not set then instead it returns the specified default value.
```

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+		return true;
+	}
+	throw new Error(
+		`Invalid value for boolean var '${varName}', got '${s}', expected 'true' or 'false'`,
+	);
+}
+
+export type HostPort = { host: string; port: number };
+/**
+ * Splits an env var in the format of `${host1}:${port1}, ${host2}:${port2}, ...` into an array of individual host/port pairings
```
Comment (dfunckt) [inline src/index.ts:102]:
```suggestion
 * Splits an env var in the format of `${host1}:${port1}, ${host2}:${port2}, ...`
 * into an array of individual host/port pairings.
```

**Comment by Page-** [comment]
This is primarily taken from https://github.com/balena-io/open-balena-api/blob/v3.5.0/src/lib/config.ts in an effort to standardize env var parsing/handling across our services

**Review by dfunckt — COMMENTED** [review]
Looks great, left a few minor -- formatting mostly -- comments. Merge at will.

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+	if (s == null || s === '') {
+		return defaultValue;
+	}
+	return s;
+}
+
+/**
+ * Checks that a string is a valid int and returns the number if so
+ */
+export const checkInt = (s: string) => {
```
Comment (dfunckt) [inline src/index.ts:41]: Helps to be explicit when reading the code, but also prevents subtle bugs from changing the implementation in a way that other return values are possible:

```suggestion
export const checkInt = (s: string): number | undefined => {
```

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+	varName: string,
+	defaultHosts?: HostPort[],
+): HostPort[] => {
+	const hostPairs = optionalVar(varName);
+	if (hostPairs == null) {
+		if (defaultHosts == null) {
+			throw new Error(`Missing environment variable: ${varName}`);
+		}
+		return defaultHosts;
+	}
```
Comment (fisehara) [inline src/index.ts:129]: Can't this be solved in the same way as L144:147

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+		// If it's 'true' enable it
+		return true;
+	}
+	if (trustProxy === 'false') {
+		// If it's 'false' disable it
+		return false;
+	}
+	if (trustProxy.includes('.') || trustProxy.includes(':')) {
+		// If it looks like an ip use as-is
+		return trustProxy;
```
Comment (fisehara) [inline src/index.ts:158]: It's a string so we may want to trim it to the IP block.

``` suggestion
return trustProxy.trim();
```

File: test/index.ts [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+	});
+});
+
+describe('optionalVar', () => {
+	const defaultValue = 'banana';
+	it('should return undefined when the env var is missing', () => {
+		expect(optionalVar(varName)).to.be.undefined;
+	});
+	it('should return undefined when the env var is empty', () => {
+		expect(optionalVar(varName)).to.be.undefined;
```
Comment (fisehara) [inline test/index.ts:42]:
``` suggestion
setVar('');
expect(optionalVar(varName)).to.be.undefined;
```

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+	s = s.trim();
+	if (/[^0-9]/.test(s)) {
+		return;
+	}
+	const i = parseInt(s, 10);
+	if (!Number.isFinite(i)) {
+		return;
+	}
+	return i;
+};
```
Comment (fisehara) [inline src/index.ts:63]: This is checkUnsignedInt negative values are resulting in undefined.
Is this intended, that a '-10' shouldn't be parsed?

File: test/index.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+	it('should throw when the env var is invalid', () => {
+		setVar('date');
+		expect(() => boolVar(varName)).to.throw;
+	});
+	it('should return the default value when the env var is missing', () => {
+		expect(boolVar(varName, defaultValue)).to.equal(defaultValue);
+	});
+	it('should return the default value when the env var is empty', () => {
+		setVar('');
+		expect(boolVar(varName, defaultValue)).to.equal(defaultValue);
```
Comment (fisehara) [inline test/index.ts:97]: This throws as defaultValue is 'banana' so not boolean

File: src/index.ts (hunk 8 above)
Comment (fisehara) [reply src/index.ts:158]: Ah I see in the test cases that multi IPs are expected input

**Review by fisehara — COMMENTED** [review]
Other than this LGTM.

File: test/index.ts (hunk 11 above)
Comment (Page-) [reply test/index.ts:97]: No it doesn't as the default value doesn't have to be a boolean, it's typed as
```ts
export function boolVar(varName: string): boolean;
export function boolVar<R>(varName: string, defaultValue: R): boolean | R;
export function boolVar<R>(varName: string, defaultValue?: R): boolean | R {
```
so if you pass in a boolean as the default then the return type will be `boolean` but if you pass in eg a string it'll be `boolean | string` and you'll be expected to handle it (I mean, you did choose to pass in a different type as the default..)

File: src/index.ts (hunk 10 above)
Comment (Page-) [reply src/index.ts]: Negatives should be fine, yes

File: src/index.ts (hunk 7 above)
Comment (Page-) [reply src/index.ts:129]: Actually L144:147 should have been being handled the same way as this, specifically the benefit of this is that we only apply the parsing to the env var and *not* the default value, so if the var does not exist we return it directly as the user passed

**Review by fisehara — APPROVED** [review]
LGTM

**Pull request merged**